modules/iothubdeviceping/src/iothubdeviceping.c

```diff
@@ -0,0 +1,738 @@
+        }
+        else
+        {
+            LogError("config->EH_HOST malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_KEY_NAME = (const char *)malloc(strlen(config->EH_KEY_NAME) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_KEY_NAME, (char *)config->EH_KEY_NAME, strlen(config->EH_KEY_NAME) + 1);
+        }
+        else
+        {
+            LogError("config->EH_KEY_NAME malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_KEY = (const char *)malloc(strlen(config->EH_KEY) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_KEY, (char *)config->EH_KEY, strlen(config->EH_KEY) + 1);
+        }
+        else
+        {
+            LogError("config->EH_KEY malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_COMP_NAME = (const char *)malloc(strlen(config->EH_COMP_NAME) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_COMP_NAME, (char *)config->EH_COMP_NAME, strlen(config->EH_COMP_NAME) + 1);
+        }
+        else
+        {
+            LogError("config->EH_COMP_NAME malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_PARTITION_NUM = (const char *)malloc(strlen(config->EH_PARTITION_NUM) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_PARTITION_NUM, (char *)config->EH_PARTITION_NUM, strlen(config->EH_PARTITION_NUM) + 1);
+        }
+        else
+        {
+            LogError("config->EH_PARTITION_NUM malloc returned NULL\r\n");
+        }
+    }
+    return clonedConfig;
+}
+
+static MODULE_HANDLE IoTHubDevicePing_Create(BROKER_HANDLE brokerHandle, const void *configuration)
+{
+    // set g_echoreply false
+    g_echoreply = false;
+    IOTHUBDEVICEPING_HANDLE_DATA *result = NULL;
+    if (
+        (brokerHandle == NULL) ||
+        (configuration == NULL) ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->DeviceConnectionString == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_HOST == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_KEY_NAME == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_KEY == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_COMP_NAME == "NULL"))
+    {
+        LogError("invalid arg brokerHandle=%p, configuration=%p\r\n", brokerHandle, configuration);
+        result = NULL;
+    }
+    else
+    {
+        result = malloc(sizeof(IOTHUBDEVICEPING_HANDLE_DATA));
+        if (result == NULL)
+        {
+            LogError("malloc returned NULL\r\n");
+        }
+        else
+        {
+            result->config = (IOTHUBDEVICEPING_CONFIG *)clone_iothub_device_ping_config(configuration);
+            result->iotHubClientHandle = IoTHubClient_LL_CreateFromConnectionString(((const IOTHUBDEVICEPING_CONFIG *)configuration)->DeviceConnectionString, HTTP_Protocol);
```

Comment:
Why aren't you using the IoT Hub module? There's a lot of code in here that duplicates the logic in that module, and I don't think we should maintain two copies of it.

I think the device_ping sample gateway should load two modules: ping and iothub. The ping module publishes a ping message to the broker, and records the time when it was published. The iothub module receives the ping message and sends it to IoT Hub. The ping module also listens for events on the backing event hub and records the timestamp when the event arrives.